plugins/polio/api/vaccines/stock_management.py

```diff
@@ -78,11 +75,169 @@ class VaccineStockManagementReadWritePerm(GenericReadWritePerm):
+            elif current_order == "vaccine_type":
+                queryset = queryset.order_by("vaccine")
+            elif current_order == "-vaccine_type":
+                queryset = queryset.order_by("-vaccine")
+
+        return queryset
+
+
 class VaccineStockManagementViewSet(ModelViewSet):
     permission_classes = [VaccineStockManagementReadWritePerm]
     serializer_class = VaccineStockSerializer
     http_method_names = ["get", "head", "options"]
 
+    model = VaccineStock
+
+    filter_backends = [SearchFilter, StockManagementCustomFilter]
+    search_fields = ["vaccine", "country__name"]
+
+    @action(detail=True, methods=["get"])
+    def usable_vials(self, request, pk=None):
+        if pk is None:
+            return Response({"error": "No VaccineStock ID provided"}, status=status.HTTP_400_BAD_REQUEST)
+
+        try:
+            vaccine_stock = self.get_queryset().get(id=pk)
+        except VaccineStock.DoesNotExist:
+            return Response({"error": "VaccineStock not found"}, status=status.HTTP_404_NOT_FOUND)
+
+        # First find the corresponding VaccineRequestForms
+        vrfs = VaccineRequestForm.objects.filter(
+            campaign__country=vaccine_stock.country, vaccine_type=vaccine_stock.vaccine
+        )
+        if not vrfs.exists():
+            return Response(
+                {"error": "No VaccineRequestForm found for the given VaccineStock Country and Vaccine Type"},
+                status=status.HTTP_404_NOT_FOUND,
+            )
+
+        # Then find the corresponding VaccineArrivalReports
+        arrival_reports = VaccineArrivalReport.objects.filter(request_form__in=vrfs)
+        if not arrival_reports.exists():
+            return Response(
+                {"error": "No VaccineArrivalReport found for the given VaccineStock Country and Vaccine Type"},
+                status=status.HTTP_404_NOT_FOUND,
+            )
+
+        # Use the doses_per_vial from the latest VaccineArrivalReport for all calculations
+        latest_arrival_report = arrival_reports.latest("arrival_report_date")
+
+        destruction_reports = DestructionReport.objects.filter(vaccine_stock=vaccine_stock).order_by(
+            "destruction_report_date"
+        )
+        incident_reports = IncidentReport.objects.filter(vaccine_stock=vaccine_stock).order_by(
+            "date_of_incident_report"
+        )
+        stock_movements = OutgoingStockMovement.objects.filter(vaccine_stock=vaccine_stock).order_by("report_date")
+
+        results = []
+
+        for report in arrival_reports:
+            results.append(
+                {
+                    "date": report.arrival_report_date,
+                    "action": "PO #" + report.po_number if report.po_number else "Stock Arrival",
+                    "vials_in": report.vials_received,
+                    "doses_in": report.doses_received,
+                    "vials_out": None,
+                    "doses_out": None,
```

Comment:
I guess the front-end has to guess if this is an "arrival", a "destruction", an "incident" or whatever?
Is this OK or should we add a `kind` field?

Reply:
Ok it makes sense, even if the frontend will not use it (yet)